### ISSUE TITLE
Faster Groovy compilation

### ISSUE THREAD

**Issue by wolfs** [gradle/gradle#6062]
Groovy compilation via Gradle (`GroovyCompile`) currently doesn't support compile avoidance or incremental compilation. When we want to have IDEA to delegate to Gradle for compilation and test execution in the Gradle build, we need at least incremental compilation for a good developer experience.

**Comment by adammurdoch** [comment]
> we need at least incremental compilation for a good developer experience.

I would expect compile avoidance to have a better return for the effort, given that there are a lot of groovy compile tasks and a lot of projects.

**Comment by lptr** [comment]
Doing compile avoidance for Groovy would also mean that we'd need to abstract out the Java compile avoidance logic. With some more work we could make it into a public API, so Kotlin could implement its own compile avoidance. I think this could be another good reason to do Groovy compile avoidance.

**Comment by wolfs** [comment]
See https://github.com/gradle/gradle/issues/5226 for supporting compile avoidance for other languages than Java.

**Comment by wolfs** [comment]
>> we need at least incremental compilation for a good developer experience.
>
> I would expect compile avoidance to have a better return for the effort, given that there are a lot of groovy compile tasks and a lot of projects

I think we should start with compile avoidance and see where we get.

A datapoint why incremental compilation would be good is that e.g. `core:compileTestGroovy` takes 40s on CI: https://e.grdev.net/s/klpldicnkwd4o/timeline?sort=duration&task=trevx2rfm4gr4&taskFilter=Groovy

**Comment by gclayburg** [comment]
Is there a chance this work could be incorporated somehow in the Apache groovy project?  I've been using GroovyScriptEngine and noticed it too has trouble with incremental compilation when static members are used.  See this issue for details: [GroovyScriptEngine doesn't always reload classes](https://issues.apache.org/jira/projects/GROOVY/issues/GROOVY-8656)

Or maybe there is a way to use gradle to provide similar functionality to what GroovyScriptEngine does now?

**Comment by lptr** [comment]
There are other ways around slow Groovy compilation:

- stop working in `:core` and other large projects, and break them into smaller chunks
- reduce the size of test source sets, so each compilation would work on a smaller number of source files
- stop working with `:core`'s existing tests, and instead copy anything we need to a new test source set directory

The last one is more of a stop-gap measure, and is probably not feasible. Modularizing tests more however could be beneficial regardless of why we do it.

I'm not saying we should go this path, though. But if we want to enable delegation without having incremental Groovy compilation (regardless if we have compile avoidance for Groovy), then we'll need some way to reduce the number of sources that need to be recompiled.

**Comment by big-guy** [comment]
Have we measured where the slowness is to see if there are any quick wins?

Could we also introduce parallel compilation within a compile task?

**Comment by adammurdoch** [comment]
The improvements to compilation don't have to be very good initially. For example, we can reuse the incremental java compilation as-is and unmodified for groovy compilation, and this will be perfectly fine for our purposes. It's not a general solution, as we're making an accuracy trade-off, so would need to be behind an opt-in until we find the time to finish it up. It's the same story for compile avoidance.

**Comment by melix** [comment]
I agree with @adammurdoch that compile avoidance is likely to have a bigger impact, at least on the Groovy build itself. If shouldn't be too hard to implement. Basically, we can reuse `@CompileClasspath`, but it's not done for a reason: the Groovy compiler has slightly different semantics than the Java one.

In particular, the Groovy compiler has AST transformations (global or local). If a transformation code is updated, we need to recompile everything. However, the fix should be pretty simple: Gradle already internally separates the "compiler" classpath (compiler plugins aka AST xforms) from the "compile classpath" (libraries). So we could apply `@CompileClasspath` on the library part, while keep using `@Classpath` on the compiler part. We just need to expose this to the user, which is not done today.

There's also a slight complication, in that AST transformations can very well access arbitrary resources. In this case I think we should just provide a way to declare those resources as inputs (they should already, really).

Eventually, Groovy has this "compiler configuration" which is itself a Groovy script. Whenever it changes, we need to rebuild. But honestly I think compile avoidance shouldn't be much effort.

Then comes incremental compilation into play. Again, the semantics are the same as Java, so I think we could reuse most of what we do for Java. From what I remember in the code this is not trivial though, because the Groovy compile tasks and the Java compile tasks are not really implemented the same way. We also have to think about stub generation, but the biggest issue is again AST transforms: because we don't know what they do, if an AST xform is found on the compiler classpath, we have to disable incremental compilation. It's extremely similar to what we have for annotation processors in Java, only Groovy allows more things.

**Comment by melix** [comment]
> I think the key here is that we aren’t needing perfection. If it can work most of the time it’s good enough since this is really just a development centric feature.

We understand this, and on the other hand we don't want to trade correctness for performance. So we have to find the right way to do it.

**Comment by wolfs** [comment]
Let's keep this epic open.
